### ISSUE TITLE
Whenever I restart a local network like hardhat-network metamask do not detect changes and keep caching old nonces

### ISSUE THREAD

**Issue by wighawag** [MetaMask/metamask-extension#12211]
**Describe the bug**
Whenever I restart a local network like hardhat-network metamask do not detect changes and keep caching old nonces.

**Steps to reproduce (REQUIRED)**

1. launch a network like hardhat via `hardhat node` (this will have chainId 31337)
2. add network to metamask where you specify http://localhost:8545 and chainId = 31337
3. make transaction via metamask on that network, this update the nonce of the account
4. repeat that for further accounts
5. stop the network and launch a new one (note that by default `hardhat node` will reuse the chainId (which is expected from a developer experience) but use a different genesisHash for block 0
6. Now you can perform another tx with metamask, but this will either silently fails or fails with an error because the nonce metamask use is the one cached from the previous network
7. In order to make metamask work with that new network, you have to click on settings -> advanced -> reset account and that for each account

**Expected behavior**
Metamask should be able to detect that the network changed (genesis Hash are different) and clear the cache when that happen
Alternatively Metamasl should provide an api to request cache clear (does not need to clear the history if the network is the same as previous)

**Comment by nickgrealy** [comment]
> 7. In order to make metamask work with that new network, you have to click on settings -> advanced -> reset account and that for each account

thank you

**Comment by gregegan** [comment]
👍 Huge pain point in development

**Comment by Xanewok** [comment]
Seems to be a dupe of #11298

**Comment by wighawag** [comment]
Any update on this @danfinlay ?
This is a huge dev experience issue

**Comment by wighawag** [comment]
A rpc method to clear the cache could be cool, to ensure an app cannot accidently delete the data from a user, that function would only clear the cache if the genesis block differs

of course if would be better if metamask handled that automatically but I assume you are not fixing this issue because it might be hard for you to fix it properly. So this could be  a nice workaround ?

**Comment by danfinlay** [comment]
Yeah, we’ve reviewed ways to fix this and they were all labor intensive, so slipped in priority. I could see allowing localhost apps the ability to request permission to reset nonce maybe.

**Comment by wighawag** [comment]
By the way, currently I fetch the genesis Hash (using "earliest")  to warn devs to clear the metamask cache if the hash is different than last time but if I use metamask provider to fetch that, metamask will cache that request too and so it does not work. I need to use the  localhost url directly which is not always ideal.

A nice thing to have could also be a way to tell metamask that I want to skip the cache. it could be per request, or a global clear cache that would only work on chainId 1337 / 31337 for example.

and this could be where metamask would check if the cache is invalid too. 

Actually thinking a bit more, all of it could be handled by the following:

```
const genesis = await ethereum.request({method: "eth_getBlockByNumber", params: ["earliest", false]});
```

When this specifi request is performed,Metamask can simply fetch it, even if it has a cache for it and compare with its cache
if it differs, it knows that whatever it cached is invalid. It would then reset the nonce, etc....

Should be simple to implement, no ?

**Comment by 0xGuybrush** [comment]
> > Yeah, we’ve reviewed ways to fix this and they were all labor intensive, so slipped in priority. I could see allowing localhost apps the ability to request permission to reset nonce maybe.
> 
> It looks like the nonce is coming from a call to [`NonceTracker.getNonceLock()`](https://github.com/MetaMask/nonce-tracker/blob/e98ef412922bbcfd5e1e63f3f9602770ba2475a3/src/NonceTracker.ts#L151-L171), is that right?
> 
> If so, it looks like it's already fetching both local and network nonce details in the `param` property, as well as the suggested `nextNonce`.
> 
> What if [`async getNextNonce(address, networkClientId)`](https://github.com/MetaMask/metamask-extension/blob/2e91b544a91bdb4f890e95e9d77bdf20f71643a5/app/scripts/metamask-controller.js#L5345-L5352), instead of doing this:
> 
> ```js
> async getNextNonce(address, networkClientId) {
>     const nonceLock = await this.txController.getNonceLock(
>       address,
>       process.env.TRANSACTION_MULTICHAIN ? networkClientId : undefined,
>     );
>     nonceLock.releaseLock();
>     return nonceLock.nextNonce;
>   }
> ```
> 
> did this:
> 
> ```js
> async getNextNonce(address, networkClientId) {
>   const nonceLock = await this.txController.getNonceLock(
>     address,
>     process.env.TRANSACTION_MULTICHAIN ? networkClientId : undefined,
>   );
>   nonceLock.releaseLock();
>   
>   return (networkClientId === 31337)
>       ? nonceLock.params.nextNetworkNonce // Anvil/Hardhat, always use network result
>       : nonceLock.nextNonce; // Use default result
> }
> ```

I'd be happy to start looking at creating/testing a PR for this if it'd be an acceptable solution? This is something that affects me daily, so it would be a real improvement for DX

**Comment by gauthierpetetin** [comment]
Solution discussed during team refinement:
- Detect when the network nonce has backtracked, and clear pending transactions automatically in that case + restrict this behaviour by chainId or by toggle (once designed by Wallet UX team)

**Comment by fvictorio** [comment]
@gauthierpetetin just doing this when the chain id is 31337 should fix 99% of the occurrences of this problem, since that's the chain id that Hardhat and Anvil use by default.

**Comment by davidmurdoch** [comment]
The transaction's nonce isn't the only issue here, MM also tracks the last block head, which would also need to be updated. We have an issue tracking this here: https://github.com/MetaMask/metamask-extension/issues/16332

**Comment by ngalee8311** [comment]
https://github.com/MetaMask/metamask-extension/issues/12211
